chips/msp432/src/uart.rs

```diff
@@ -0,0 +1,278 @@
+    BaudFraction { frac: 0.2224, reg_val: 0x22 },
+    BaudFraction { frac: 0.2503, reg_val: 0x44 },
+    BaudFraction { frac: 0.3000, reg_val: 0x25 },
+    BaudFraction { frac: 0.3335, reg_val: 0x49 },
+    BaudFraction { frac: 0.3575, reg_val: 0x4A },
+    BaudFraction { frac: 0.3753, reg_val: 0x52 },
+    BaudFraction { frac: 0.4003, reg_val: 0x92 },
+    BaudFraction { frac: 0.4286, reg_val: 0x53 },
+    BaudFraction { frac: 0.4378, reg_val: 0x55 },
+    BaudFraction { frac: 0.5002, reg_val: 0xAA },
+    BaudFraction { frac: 0.5715, reg_val: 0x6B },
+    BaudFraction { frac: 0.6003, reg_val: 0xAD },
+    BaudFraction { frac: 0.6254, reg_val: 0xB5 },
+    BaudFraction { frac: 0.6432, reg_val: 0xB6 },
+    BaudFraction { frac: 0.6667, reg_val: 0xD6 },
+    BaudFraction { frac: 0.7001, reg_val: 0xB7 },
+    BaudFraction { frac: 0.7147, reg_val: 0xBB },
+    BaudFraction { frac: 0.7503, reg_val: 0xDD },
+    BaudFraction { frac: 0.7861, reg_val: 0xED },
+    BaudFraction { frac: 0.8004, reg_val: 0xEE },
+    BaudFraction { frac: 0.8333, reg_val: 0xBF },
+    BaudFraction { frac: 0.8464, reg_val: 0xDF },
+    BaudFraction { frac: 0.8572, reg_val: 0xEF },
+    BaudFraction { frac: 0.8751, reg_val: 0xF7 },
+    BaudFraction { frac: 0.9004, reg_val: 0xFB },
+    BaudFraction { frac: 0.9170, reg_val: 0xFD },
+    BaudFraction { frac: 0.9288, reg_val: 0xFE },
+];
+
+pub struct Uart<'a> {
+    registers: StaticRef<UsciARegisters>,
+    clock_frequency: u32,
+    tx_client: OptionalCell<&'a dyn hil::uart::TransmitClient>,
+    rx_client: OptionalCell<&'a dyn hil::uart::ReceiveClient>,
+
+    tx_buffer: TakeCell<'static, [u8]>,
+    tx_len: Cell<usize>,
+    tx_index: Cell<usize>,
+}
+
+impl<'a> Uart<'a> {
+    pub(crate) const fn new(regs: StaticRef<UsciARegisters>) -> Uart<'a> {
+        Uart {
+            registers: regs,
+            clock_frequency: DEFAULT_CLOCK_FREQ_HZ,
+            tx_client: OptionalCell::empty(),
+            rx_client: OptionalCell::empty(),
+            tx_buffer: TakeCell::empty(),
+            tx_len: Cell::new(0),
+            tx_index: Cell::new(0),
+        }
+    }
+
+    fn enable_tx_interrupt(&self) {
+        self.registers.ie.modify(usci::UCAxIE::UCTXIE::SET);
+    }
+
+    fn disable_tx_interrupt(&self) {
+        self.registers
+            .ifg
+            .modify(usci::UCAxIFG::UCTXIFG::CLEAR + usci::UCAxIFG::UCTXCPTIFG::CLEAR);
+        self.registers.ie.modify(usci::UCAxIE::UCTXIE::CLEAR);
+    }
+
+    fn write_byte(&self, data: u8) {
+        self.registers.txbuf.set(data as u16);
+    }
+
+    fn start_transmit(&self) {
+        if self.tx_index.get() >= self.tx_len.get() {
+            // nothing to transmit
+            return;
+        }
+        self.enable_tx_interrupt();
+        self.transmit();
+    }
+
+    fn transmit(&self) {
+        let idx = self.tx_index.get();
+
+        self.tx_buffer.map(|tx_buf| self.write_byte(tx_buf[idx]));
+        self.tx_index.set(idx + 1);
+    }
```

Comment:
It looks like this is (asynchronously, good) transmitting one byte at a time. That's going to create a ton of interrupts. Is there no easy way to pass a length for the buffer to the uart peripheral?

Reply:
This UART-module unfortunately doesn't have any kind of FIFO, it can only transmit one byte at a time. It could use a DMA-channel for avoiding these interrupts, but I thought for the first implementation I just go the easier way without DMA.

Reply:
I think that's reasonable as a start for basic functionality.

DMA should be on the shortlist of next implementation priorities once this is merged, however.